src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs

```diff
@@ -2606,6 +2811,52 @@ internal static extern
                                     ref JOBOBJECT_BASIC_PROCESS_ID_LIST lpJobObjectInfo,
                                     int cbJobObjectLength, IntPtr lpReturnLength);
 
+        internal static bool QueryServiceConfig(NakedWin32Handle hService, out IntPtr structurePointer)
+        {
+            IntPtr lpBuffer = IntPtr.Zero;
+            DWORD bufferSize, bufferSizeNeeded = 0;
+            bool status = NativeMethods.QueryServiceConfigW(
+                hSCManager: hService,
+                lpServiceConfig: lpBuffer,
+                cbBufSize: 0,
+                pcbBytesNeeded: out bufferSizeNeeded);
+
+            lpBuffer = Marshal.AllocCoTaskMem((int)bufferSizeNeeded);
+            bufferSize = bufferSizeNeeded;
+
+            status = NativeMethods.QueryServiceConfigW(
+                hService,
+                lpBuffer,
+                bufferSize,
+                out bufferSizeNeeded);
+            structurePointer = lpBuffer;
+            return status;
+        }
+
+        internal static bool QueryServiceConfig2(NakedWin32Handle hService, DWORD infolevel, out IntPtr structurePointer)
+        {
+            IntPtr lpBuffer = IntPtr.Zero;
+            DWORD bufferSize, bufferSizeNeeded = 0;
+            bool status = NativeMethods.QueryServiceConfig2W(
+                hService: hService,
+                dwInfoLevel: infolevel,
+                lpBuffer: lpBuffer,
+                cbBufSize: 0,
+                pcbBytesNeeded: out bufferSizeNeeded);
+
+            lpBuffer = Marshal.AllocCoTaskMem((int)bufferSizeNeeded);
```

Comment:
A check should be made to confirm that status == ERROR_INSUFFICIENT_BUFFER before proceeding.

https://msdn.microsoft.com/en-us/library/windows/desktop/ms684935.aspx

Reply:
Interesting, what is best practice? Could we exclude second call by pre-allocate memory buffer?

Reply:
This technique is a common pattern in Windows APIs. It is designed to be called twice with the first call returning the buffer size needed so the correct buffer size can be allocated and supplied on the second call. You can do it all in one call, but you have to either guess how big the buffer should be or allocate a max size buffer. If you guess wrong, you'll need to resize it and call the method again anyway. A max size buffer will work, but will incur unnecessary memory allocation. With that in mind, it is optimal to call it twice and follow the expected pattern.

Reply:
I can't find this. I think your comment might be on an outdated version of the code. I made the check [here][0] and [here][1]

[0]: https://github.com/PowerShell/PowerShell/pull/4907/files#diff-5ee347b353d729b7e743da3c535d1a21R2835
[1]: https://github.com/PowerShell/PowerShell/pull/4907/files#diff-5ee347b353d729b7e743da3c535d1a21R2863

Reply:
Agreed. Steve and I made the same review comment around the same time. This has been addressed.